torch/csrc/jit/export.cpp

```diff
@@ -181,7 +181,28 @@ void encodeGraph(onnx::GraphProto * p_g, const std::shared_ptr<Graph> & g, const
   JIT_ASSERT(p_g != nullptr);
   p_g->set_name("torch-jit-export");
 
+  // The jit tracer may record some unnecessary information.
+  // We should remove the unused input and initializer.
+  // Gather all the names of useful nodes and filter out unused ones.
+  std::unordered_set<std::string> useful_names;
+  std::vector<std::string> input_names;
+  for (auto node : g->nodes()) {
+    for(auto input : node->inputs()) {
+      useful_names.insert(value_name(input));
+    }
+    for(auto output : node->outputs()) {
+      useful_names.insert(value_name(output));
+    }
+  }
+  for (auto output : g->outputs()) {
+    useful_names.insert(value_name(output));
+  }
+
   for (auto input : g->inputs()) {
+    input_names.push_back(value_name(input));
+    if (useful_names.find(value_name(input)) == useful_names.end()) {
+      continue;
```

Comment:
I was going to complain that doing this filtering processes destroys the correspondence between `module.state_dict()` and ONNX parameters, but apparently we already lost this property when @apaszke did e6cbe84bf679111316590efb4fd31ba195867a57. @apaszke, was destroying this correspondence intentional?

I will admit we have not been advertising the correspondence between `state_dict` and ONNX parameters, but if we don't have names for input parameters I am not sure how you are going to be able to line them up at all when they don't match.

Reply:
Actually, here, we have to assume the correspondence between modul.state_dict() and ONNX initializers.

Reply:
Actually, can we get rid of that correspondence and instead use pointers matching (Tensor*) or some unique id of Tensors? It seems very very fragile and will break for sure

Reply:
@dzhulgakov I think this is an orthogonal issue.

The basic problem is, suppose you have a model (subclass `torch.autograd.Module`) in PyTorch. You know what the arity of the model is (e.g., it takes one input.) You run an ONNX export. Now you have a graph which takes 1 + some big pile of parameters which are not named in any way. How do you figure out what these parameters mean? Since we don't have names on them, you only know what they are if you can construct a correspondence between each PyTorch weight and the position in it. The way we implemented it previously was to use the ordering of `state_dict()` (which is an ordered dict) to do it.

Here's another way to solve it: we could add "external labels" to inputs; these are meaningful strings that someone executing an ONNX graph can make use of. The key is to NOT have this label coincide with the internal name we refer to the input with. So, for example, instead of `graph %0, %1 { ... }` we have `graph (input: %0, weights: %1) { ... }` and you still refer to input as `%0` inside. That will solve the problem that torpedoed https://github.com/pytorch/pytorch/pull/2697